### PR TITLE
[SPARK] Add support for Kafka datasets when Spark is in its streaming configuration

### PR DESCRIPTION
**Pull request by d-m-h**
# Problem

Currently, when you read from and write to Apache Kafka using Spark streaming, the OpenLineage events that are generated lack inputs and outputs.

# Solution

This PR aims to add support for Kafka sources and sinks when Spark is running in its streaming configuration

### CODE REVIEW

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/OpenLineageRunEventBuilder.java [hunk 1]
```diff
@@ -371,8 +371,7 @@ private List<InputDataset> buildInputDatasets(List<Object> nodes) {
                     .getQueryExecution()
                     .map(
                         qe ->
-                            ScalaConversionUtils.fromSeq(qe.optimizedPlan().map(inputVisitor))
-                                .stream()
+                            ScalaConversionUtils.fromSeq(qe.analyzed().map(inputVisitor)).stream()
```
[the remark this answers is not in the document]
Comment (d-m-h) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/OpenLineageRunEventBuilder.java]: I think I managed to find a workaround. However, there is a fairly significant caveat when using a Kafka sink. Spark allows users to specify a "static" topic via `.option("topic", "<destination-topic")`, or they can do _dynamic topics_ by putting the topic name into the data frame with a column name of `topic`. 

I'll add this to the PR description.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/plan/WriteToDataSourceV2Visitor.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    BatchWrite batchWrite = write.batchWrite();
+    if (batchWrite instanceof MicroBatchWrite) {
+      MicroBatchWrite microBatchWrite = (MicroBatchWrite) batchWrite;
+      StreamingWrite streamingWrite = microBatchWrite.writeSupport();
+      Class<? extends StreamingWrite> streamingWriteClass = streamingWrite.getClass();
+      String streamingWriteClassName = streamingWriteClass.getCanonicalName();
+      if (KAFKA_STREAMING_WRITE_CLASS_NAME.equals(streamingWriteClassName)) {
+        return handleKafkaStreamingWrite(streamingWrite, write);
+      }
+    }
```
Comment (pawel-big-lebowski) [inline integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/plan/WriteToDataSourceV2Visitor.java:61]: would it make sense to log if it is not `MicroBatchWrite`?

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/plan/KafkaMicroBatchStreamStrategy.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      }
+    } else {
+      return Collections.emptySet();
+    }
+  }
+
+  private List<InputDataset> generateInputDatasets(
+      String bootstrapServersConfig, Collection<String> topics) {
+    String[] bootstrapServers = bootstrapServersConfig.split(",");
+    String namespace = "kafka://" + bootstrapServers[0];
```
Comment (pawel-big-lebowski) [inline integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/plan/KafkaMicroBatchStreamStrategy.java:83]: Isn't this different behaviour than you defined for output within this PR? 

General issue for this problem is: https://github.com/OpenLineage/OpenLineage/issues/560

**Review by pawel-big-lebowski — COMMENTED** [review]
Keeping my fingers crossed to have this working 🤞 
I put some minor comments in the code.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/plan/WriteToDataSourceV2Visitor.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+  @Slf4j
+  private static final class KafkaStreamWriteProxy {
+    private final StreamingWrite streamingWrite;
+
+    public KafkaStreamWriteProxy(StreamingWrite streamingWrite) {
+      String incomingClassName = streamingWrite.getClass().getCanonicalName();
+      if (!KAFKA_STREAMING_WRITE_CLASS_NAME.equals(incomingClassName)) {
+        throw new IllegalArgumentException(
+            "Expected the supplied argument to be of type '"
```
[the remark this answers is not in the document]
Comment (d-m-h) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/plan/WriteToDataSourceV2Visitor.java:100]: Could you expand on what you mean by this? Is the issue with the wording? Would:
```
Expected the supplied argument to have type ${KAFKA_STREAMING_WRITE_CLASS_NAME}, but received ${ACTUAL_CLASS_NAME} instead
```

[the remark this answers is not in the document]
Comment (pawel-big-lebowski) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/plan/WriteToDataSourceV2Visitor.java:108]: Just wording. If `incomingClassName` extends `KAFKA_STREAMING_WRITE_CLASS_NAME` then `equal` returns false, but `incomingClassName` is of type `KAFKA_STREAMING_WRITE_CLASS_NAME`.

**Comment by Imbruced** [comment]
@d-m-h Hi, do you plan to work on this any time soon ? Can I help you with finishing this one ?

**Comment by mobuchowski** [comment]
@d-m-h closing, because @Imbruced's PR solved that already.

**Pull request closed without merging**